### PR TITLE
DT-6533 Improved navigator and local storage cooperation

### PR DESCRIPTION
**Pull request by partisaani**
## Proposed Changes
Navigator is now more context aware:
  - If `from`, `to`, `time`, `arriveBy` and `hash` parameters are identical to those of the stored journey
    - Traversing to `ItineraryPage` detail view will automatically trigger the Navigator journey.
  - If `from`, `to`, `time` and `arriveBy` parameters are identical to those of the stored journey
    - Traversing to `ItineraryPage` list view does not necessarily purge the stored itinerary from memory.
      - Local storage entry is purged if user navigates to list view from active navigator journey either by using the arrow icon or by browsing back using browser history
    - Traversing to `ItineraryPage` detail view does not purge the stored itinerary from memory, but a Navigator journey is not triggered
  - If any of the parameters is different
    - Traversing to `ItineraryPage` list view or detail view will purge the stored Navigator journey from local storage

## Pull Request Check List

  - [x] A reasonable set of unit tests is included
  - [x] Console does not show new warnings/errors
  - [x] Changes are documented or they are self explanatory
  - [x] This pull request does not have any merge conflicts
  - [x] All existing tests pass in CI build

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

### CODE REVIEW

File: app/component/itinerary/ItineraryDetails.js
Comment (partisaani) [inline app/component/itinerary/ItineraryDetails.js:1]: Whitespace and format changes only

File: app/component/itinerary/ItineraryPage.js [hunk 1]
```diff
@@ -745,16 +739,63 @@ export default function ItineraryPage(props, context) {
...
   }
 
+  /**
+   * Checks if itinerary in local storage was set with identical parameters.
+   *
+   * @param {{from: string, to: string, time: number, arriveBy: boolean, hash: number}} storedParams
+   * @param {boolean} shouldCompareHash false bypasses hash check when user navigates to ItineraryPage root with the list of suggested itineraries
+   * @returns true if stored itinerary was set with identical parameters
+   */
+  const isMatchingParams = (storedParams, shouldCompareHash) => {
```
Comment (Antiik91) [inline app/component/itinerary/ItineraryPage.js]: I think these should be moved in some util file if possible, in order not to bloat the ItineraryPage more than necessary.

**Review by Antiik91 — APPROVED** [review]

File: app/component/itinerary/ItineraryPageUtils.js [hunk 2]
```diff
@@ -535,3 +536,50 @@ export function quitIteration(plan, newPlan, planParams, startTime) {
...
+export const storedParamsMatchURL = (storedParams, match) => {
+  if (!storedParams || !match) {
+    return false;
+  }
+
+  return (
+    storedParams.from === match?.params?.from &&
+    storedParams.to === match?.params?.to &&
+    storedParams.time === match?.location?.query?.time &&
+    storedParams.arriveBy === match?.location?.query?.arriveBy
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPageUtils.js]: no need for match?, it is tested in the beginning of the function. Actually, can match be undefined?

File: app/component/itinerary/StartNavi.js [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
+  const handleClick = () => {
+    setNavigation(true);
+    setLatestNavigatorItinerary({
+      itinerary,
+      params: {
+        from: match.params.from,
+        to: match.params.to,
+        arriveBy: match.location.query.arriveBy,
+        time: match.location.query.time,
+        index: match.params.hash,
```
Comment (vesameskanen) [inline app/component/itinerary/StartNavi.js]: Unfortunately, we have to deal with secondaryHash as well.

File: app/component/itinerary/ItineraryPage.js [hunk 4]
```diff
@@ -750,11 +746,13 @@ export default function ItineraryPage(props, context) {
...
-    setNavigation(
-      storedItinerary?.end && Date.parse(storedItinerary.end) > Date.now(),
-    );
+    const { params: storedParams } = getLatestNavigatorItinerary();
+    if (
+      (hash === undefined || hash === null) &&
+      !storedParamsMatchURL(storedParams, match)
+    ) {
+      clearLatestNavigatorItinerary();
+    }
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js]: I think you need something like this here:
 ```
 if (isStoredItineraryRelevant(match)) { 
    setnavigation(true);
} else {
  clearLatestNavigatorItinerary();
}
```
relevance testing can handle all  hash checking logic

File: app/component/itinerary/ItineraryPage.js [hunk 5]
```diff
@@ -776,6 +774,9 @@ export default function ItineraryPage(props, context) {
       makeRelaxedQuery();
       makeRelaxedScooterQuery();
     }
+    if (isStoredItineraryRelevant(match)) {
+      setNavigation(true);
+    }
   }, [
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js]: I don't think this is necessary. Itinerary search parameter change never triggers navi mode.

File: app/component/itinerary/ItineraryPage.js [hunk 6]
```diff
@@ -787,6 +788,9 @@ export default function ItineraryPage(props, context) {
   useEffect(() => {
     navigateMap();
     setMapState({ center: undefined, zoom: undefined, bounds: undefined });
+    if (isStoredItineraryRelevant(match)) {
+      setNavigation(true);
+    }
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js]: Not necessary. Hash change should not trigger navi mode.

**Review by vesameskanen — CHANGES_REQUESTED** [review]
Note that when you open itinerary detaul view of a  bike and transit itinerary, hash is not itinerary index. It can be found from secondaryHash.

As far as I can see, we need to 'auto relaunch' navigation only in one occasion: page reloading. This is handled in useEffect [].

File: app/component/itinerary/StartNavi.js (hunk 3 above)
Comment (partisaani) [reply app/component/itinerary/StartNavi.js]: Will add

File: app/component/itinerary/ItineraryPageUtils.js (hunk 2 above)
Comment (partisaani) [reply app/component/itinerary/ItineraryPageUtils.js]: It is likely match cannot be undefined, but I don't think being prepared for such occasion is detrimental

File: app/component/itinerary/ItineraryPage.js (hunk 6 above)
Comment (partisaani) [reply app/component/itinerary/ItineraryPage.js]: Will remove

File: app/component/itinerary/ItineraryPage.js (hunk 5 above)
Comment (partisaani) [reply app/component/itinerary/ItineraryPage.js]: Will remove

**Comment by partisaani** [comment]
> As far as I can see, we need to 'auto relaunch' navigation only in one occasion: page reloading. This is handled in useEffect [].

Will refactor with this in mind. For possible future reference the current implementation enables the user to browse between itineraries and their detail views. Browsing to the exact same search result auto-launches Navigator.

Only explicit backtrack from an initiated journey or a change in `from`, `to`, `time` or `arriveBy` params signal to purge the stored itinerary.

File: app/component/itinerary/ItineraryPageUtils.js [hunk 7]
```diff
@@ -535,3 +536,52 @@ export function quitIteration(plan, newPlan, planParams, startTime) {
...
+  } = match;
+
+  return (
+    (itinerary &&
+      Date.parse(itinerary.end) > Date.now() &&
+      storedParamsMatchURL(params, match) &&
+      params?.index == secondHash) ||
+    params?.index == hash
+  );
+};
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPageUtils.js:563]: - last condition is apparently wrong, shoudl be ( params.index == secondHash || params.index == hash )
- would it be easier and safer to save hash and secondHash and compare both ?
- let code crash if match params is not defined. unnecessary validation to keep code running  although everyhing is all wrong is not wise

**Review by vesameskanen — CHANGES_REQUESTED** [review]

File: app/component/itinerary/ItineraryPageUtils.js (hunk 7 above)
Comment (partisaani) [reply app/component/itinerary/ItineraryPageUtils.js:563]: Refactored and fixed.

**Review by vesameskanen — APPROVED** [review]
very good

**Pull request merged**